Return highlighted excerpt of result

openspectrum opened the pull request:
Addresses issue #100.

If configuration is set to

```ruby
:using => {
  :tsearch => { :highlight => true }
}
```

then queries will return records with an additional `pg_highlight` attribute.

Currently uses the defaults set by the Postgres function ts_headline, as specified here: http://www.postgresql.org/docs/current/static/textsearch-controls.html#TEXTSEARCH-HEADLINE

Also supports custom delimiters

```ruby
:using => {
  :tsearch => { 
    :highlight => { 
      :start_sel => "begin", 
      :stop_sel => "end" 
    }
  }
}
```

will override the default `<b>` and `</b>` query delimiters in the highlighted results.

NOTE: the pg function used (`ts_headline()`) is _not_ supported on pg versions < 9.0. I know that the gem is intended to support 8.0 and above, but this seemed like a useful enough feature get started. Any way to work around?

openspectrum commented:
So the build is failing because Rubocop wants class definitions to have <= 100 lines of code, and [/lib/pg_search/features/tsearch.rb](https://github.com/openspectrum/pg_search/blob/tw/pg_highlight/lib/pg_search/features/tsearch.rb) is 107. :/

I could resolve by either:

1. Extracting into a module
- Changing the Rubocop config

Do you have a preference @nertzy?

nertzy commented:
It would be better to refactor some of the methods into service objects. For instance, tsquery_for_term is way too complicated.

But for now it would be OK with me to add a Rubocop comment that turns off the file size check for just tsearch.rb until we can get a good refactor in. Just add a line to the top of the file that turns off that rule:

```ruby
# rubocop:disable Metrics/ClassLength
```

marshall-lee commented:
This is awesome!

openspectrum commented:
Thanks, @nertzy. I'd rather do it right the first time, so I'll refactor them out as suggested.

openspectrum commented:
Ok, so here's the deal. I cleaned up this PR, added documentation, and added a workaround to appease Rubocop.

Then, in a [separate branch](https://github.com/openspectrum/pg_search/tree/tw/refactor-tsearch), I did some serious refactoring.

I didn't want to bundle all of it into this one PR, so we can do it in two phases:

1. If this PR looks good, then you can merge and the new feature will be available.
- I'll make another PR with the refactorings so that the code is actually a little cleaner (not perfect, but better).

What do you think, @nertzy? Will that work for you?

nertzy commented:
I'm sorry that I've taken so long to address this pull request. Unfortunately, the new subquery structure in pg_search 1.0 breaks this pull request. I spent a few minutes trying to refactor what you've built to act similarly to how `.with_pg_search_rank` works, but to no avail.

So I'm going to close this for now. I'd gladly accept a pull request based off of this that adapts things to the new subquery, and that makes the highlight optional through something like a `.with_pg_search_highlight` method chained onto the scope.

tannerwelsh commented:
Hey @nertzy no worries, I got pulled in other directions too. Sadly this is not a high priority for me anymore, so don't think I'll be able to work on the feature anytime soon.

Thanks for the helpful suggestions and feedback earlier though! Hope that #280 is able to make it through by building on this one. :smile:

olivierlacan commented:
@nertzy Was documentation for this feature merged into the master branch README by mistake? I keep trying to use `with_pg_search_highlight` on a search scope and the method always comes up as undefined.

olivierlacan commented:
By the way I'm referring to https://github.com/Casecommons/pg_search/blob/ff1af346322ca54808b3f53081149b8cac55bb76/README.md#highlight-postgresql-90-and-newer-only

The recommended configuration causes the following error:

```
ArgumentError: Unknown key: :start_sel.
```

nertzy commented:
It's not an error. It's merged to master but not yet in a released gem version, I believe. I'll try to get a new version cut very soon, sorry for the confusion!

olivierlacan commented:
Oh, fantastic. Thank you.

sintro commented:
There are still error in documentation (the sample content is:
```
:using => {
  :tsearch => { 
      :start_sel => "begin", 
      :stop_sel => "end" 
  }
}
```
but not 
```
:using => {
  :tsearch => { 
    :highlight => { 
      :start_sel => "begin", 
      :stop_sel => "end" 
    }
  }
}
```
)
Also I found that there is error when trying to use `with_pg_search_highlight` method in combination with `associated_against`, the error description is something like
```
table "pg_search_3be97ef13e109a88624534" is absent in FROM-clause 
```

heisters commented:
@sintro I fixed the documentation bug in #312
